src/drivers/imu/analog_devices/adis16507/ADIS16507.cpp

```diff
@@ -88,96 +91,147 @@ void ADIS16507::exit_and_cleanup()
-	const uint16_t FIRM_REV = RegisterRead(Register::FIRM_REV);
-	const uint16_t FIRM_DM = RegisterRead(Register::FIRM_DM);
-	const uint16_t FIRM_Y = RegisterRead(Register::FIRM_Y);
+	const uint16_t serial = RegisterRead(Register::SERIAL_NUM);
+	const uint16_t rev = RegisterRead(Register::FIRM_REV);
+	const uint16_t daymonth = RegisterRead(Register::FIRM_DM);
+	const uint16_t year = RegisterRead(Register::FIRM_Y);
 
-	PX4_INFO("Serial Number: 0x%X, Firmware revision: 0x%X Date: Y %X DM %X", SERIAL_NUM, FIRM_REV, FIRM_Y, FIRM_DM);
+	PX4_INFO("Serial Number: 0x%X, Firmware revision: 0x%X Date: Y %X DM %X", serial, rev, year, daymonth);
 
 	return PX4_OK;
 }
 
+bool ADIS16507::Configure()
+{
+	struct {
+		uint16_t reg;
+		uint16_t val;
+	} defaults[] = {
+		// Default 0x00C1
+		// - Change Data Ready polarity to active low
+		{ Register::MSC_CTRL, 0x00C0 },
+	};
+
+	// Write default register configuration
+	for (const auto &r : defaults) {
+		RegisterWrite(r.reg, r.val);
+	}
+
+	// We must wait for changes to apply
+	px4_usleep(5000);
+
+	// Check that all are configured
+	for (const auto &r : defaults) {
+		if (!RegisterCheck(r.reg, r.val)) {
+			return false;
+		}
+	}
+
+	// Accelerometer only has a single measurement range and scale
+	_px4_accel.set_range(392);
+	_px4_accel.set_scale(392.f / 32'000.f); // 32,000 -> 392 m/sec^2
+
+	// Check gyroscope measurement range
+	const uint16_t rang_mdl = RegisterRead(Register::RANG_MDL);
+
+	// sanity check RANG_MDL [1:0] Reserved, binary value = 11
+	if (rang_mdl & (Bit1 | Bit0)) {
+		const uint16_t gyro_range = (rang_mdl & (Bit3 | Bit2)) >> 2;
+
+		if (gyro_range == 0b11) {
+			PX4_DEBUG("Gyro Range ±2000°/sec");
+			// 11 = ±2000°/sec (ADIS16507-3BMLZ)
+			_px4_gyro.set_range(math::radians(2000.f));
+			_px4_gyro.set_scale(math::radians(1.f / 10.f)); // scaling 10 LSB/°/sec -> rad/s per LSB
+
+		} else if (gyro_range == 0b01) {
+			PX4_DEBUG("Gyro Range ±500°/sec");
+			// 01 = ±500°/sec (ADIS16507-2BMLZ)
+			_px4_gyro.set_range(math::radians(500.f));
+			_px4_gyro.set_scale(math::radians(1.f / 40.f)); // scaling 40 LSB/°/sec -> rad/s per LSB
+
+		} else if (gyro_range == 0b00) {
+			PX4_DEBUG("Gyro Range ±125°/sec");
+			// 00 = ±125°/sec (ADIS16507-1BMLZ)
+			_px4_gyro.set_range(math::radians(125.f));
+			_px4_gyro.set_scale(math::radians(1.f / 160.f)); // scaling 160 LSB/°/sec -> rad/s per LSB
+		}
+	}
+
+	return true;
+}
+
 void ADIS16507::RunImpl()
 {
 	const hrt_abstime now = hrt_absolute_time();
 
 	switch (_state) {
 	case STATE::RESET:
+		PX4_DEBUG("Resetting");
 		perf_count(_reset_perf);
-		// GLOB_CMD: software reset
-		RegisterWrite(Register::GLOB_CMD, GLOB_CMD_BIT::Software_reset);
-		_reset_timestamp = now;
+
+#ifdef GPIO_ADIS16507_RESET
+		PX4_DEBUG("Hardware reset");
+		GPIO_ADIS16507_RESET(1);
+		// Minimum 10us
+		px4_usleep(15);
+		GPIO_ADIS16507_RESET(0);
+#else
+		PX4_DEBUG("Software reset");
+		RegisterWrite(Register::GLOB_CMD, Register::GLOB_CMD_BIT::Software_reset);
```

Comment:
You could probably just do the software reset regardless (first). If the GPIO reset worked it won't matter.